### PR TITLE
📖 Add docs: custom markers for unsupported file extensions

### LINKED ISSUE #4829: Add docs: how to create custom markers for unsupported file extensions
### What do you want to happen?

We should add a documentation page that shows how to use **Kubebuilder as a library** to define your **own marker support** for file types not handled by core (like `.rs`, `.tpl`, `.java`, etc.).

This is useful when:
- You're building an **external plugin**
- You want to scaffold files that are not Go
- Your file extensions aren't (and shouldn't be) part of the core `commentsByExt` map

The doc should show:
- How to create your own `Marker` using `machinery.Marker`
- How to plug that into your scaffolding logic

See this example with `.rs`:  
https://github.com/kubernetes-sigs/kubebuilder/pull/4827#pullrequestreview-2866897777

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by nerdeveloper**
## Summary

- Adds documentation on creating custom markers for unsupported file extensions
- Shows how to use Kubebuilder as a library for external plugins  
- Includes examples for Rust, Java, Python, and template files

## Related

- Fixes #4829
- Context from PR #4827

## Test Plan

- [x] Documentation builds correctly
- [x] Code examples tested and compile without errors
- [x] Added to SUMMARY.md table of contents

### CODE REVIEW

File: docs/book/src/plugins/extending/custom-markers.md [hunk 1]
```diff
@@ -0,0 +1,349 @@
+# Creating Custom Markers for Unsupported File Extensions
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md:1]: I think the documentation could be How to create Custom Markers.
For those who use Kubebuilder as a library, it would be possible to create their own custom markers, regardless of whether it is for an extension valid or not for Golang projects.

Comment (camilamacedo86) [reply docs/book/src/plugins/extending/custom-markers.md:1]: Can we make it more generic?
Also, is it possible to do it as low verbose as possible?

File: docs/book/src/plugins/extending/custom-markers.md [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+When building external plugins for Kubebuilder, you may need to scaffold files with extensions that aren't natively supported by Kubebuilder's marker system (such as `.rs` for Rust, `.java` for Java, `.tpl` for templates, etc.). This guide shows you how to use Kubebuilder as a library to create your own custom marker support for any file extension.
+
+## When to Use Custom Markers
+
+Custom markers are useful when:
+
+- You're building an **external plugin** for a language not natively supported by Kubebuilder
+- You want to scaffold files with custom extensions (`.rs`, `.java`, `.tpl`, `.py`, etc.)
+- Your file extensions aren't (and shouldn't be) part of the core `commentsByExt` map
+- You need to maintain scaffolding markers in non-Go files
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md]: I think the documentation could be How to create Custom Markers.
For those who use Kubebuilder as a library, it would be possible to create their own custom markers, regardless of whether it is for an extension valid or not for Golang projects. 

So, it would not only be valid for external references

File: docs/book/src/plugins/extending/custom-markers.md [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+    var builder strings.Builder
+    if !strings.HasPrefix(trimmed, "+") {
+        builder.WriteString("+")
+    }
+    builder.WriteString(trimmed)
+    if !strings.HasSuffix(trimmed, ":") {
+        builder.WriteString(":")
+    }
+    return builder.String()
+}
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md]: Could you let me know if you tested this example?
Can it work without any change in the code at this moment? 
Is the code covered with tests to ensure that this example would be possible?

File: docs/book/src/plugins/extending/custom-markers.md [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+```
+
+## Complete Example: Rust Plugin with Custom Markers
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md]: Do we need all those examples?
Could we not have only 1 example?
How could we simplify the doc and its maintainability? 
Just to share what is required

File: docs/book/src/plugins/extending/custom-markers.md [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+2. **Handle Comments Correctly**: Different languages have different comment syntax. Make sure to map the correct comment style for each file extension.
+
+3. **Error Handling**: Always validate file extensions and return clear error messages when unsupported files are encountered.
+
+4. **Maintain Compatibility**: When possible, follow the same patterns as Kubebuilder's core marker system to maintain consistency.
+
+5. **Document Your Markers**: Clearly document what markers your plugin supports and where they should be placed.
+
+6. **Testing**: Test your marker implementation with various file types and edge cases.
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md]: Best Practices is something subjective, I think it's better to avoid this term.

File: docs/book/src/plugins/extending/custom-markers.md [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        }
+    }
+
+    // Write back to file
+    return afero.WriteFile(s.fs, "src/main.rs",
+        []byte(strings.Join(lines, "\n")), 0644)
+}
+```
+
+## Conclusion
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md]: Could you please review the other documentation we have? 
We do not use a conclusion. We should always follow the same layout and standards.

**Comment by nerdeveloper** [comment]
Hi @camilamacedo86,

Thank you for the review feedback! I've updated the documentation to address your comments:

✅ Changed title to "Creating Custom Markers" (broader scope, not just for unsupported extensions)
✅ Simplified to only ONE example (Rust) instead of multiple examples
✅ Removed "Best Practices" section (subjective term)
✅ Removed "Conclusion" section (to match other docs)
✅ Made the documentation more concise and maintainable

Regarding testing: I performed minimal testing of the marker generation logic by creating a standalone Go program that tests the core marker creation code. The test confirms:
- The code compiles without errors
- Markers are generated with correct comment syntax for different file types (e.g., `// +rust:scaffold:imports` for Rust, `# +python:scaffold:functions` for Python)
- The marker format is correct

However, I have NOT tested the complete integration with external plugins end-to-end (the full example with templates and plugin integration).

**Question:** What would you prefer for this documentation?
1. Keep the current example as a pattern/reference (with a note that it's not fully integration-tested)
2. Simplify further to only show the minimal testable code (just the marker creation)
3. Add a working example to an external test repository and link to it

Please let me know your preference, and I'll update accordingly!

File: docs/book/src/plugins/extending/custom-markers.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+## Key Considerations
+
+1. **Unique Prefixes**: Choose a unique prefix for your plugin to avoid conflicts (e.g., `+rust:scaffold:`, `+java:scaffold:`)
+
+2. **Comment Syntax**: Different languages have different comment syntax. Ensure you map the correct comment style for each file extension
+
+3. **Error Handling**: Validate file extensions and return clear error messages for unsupported files
+
+4. **Testing**: Test your marker implementation with various scenarios to ensure reliability
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md]: I think all is great !!!
Well done 👍 Good job but wdyt about we remove this part? 
I think is best we avoid adding things that are more opinionated and keep it as low verbose as possible.

File: docs/book/src/plugins/extending/custom-markers.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+Example of a marker in a Go file:
+```go
+// +kubebuilder:scaffold:imports
+```
+
+## Implementation Example
+
+Here's how to implement custom markers for Rust files (`.rs`). This same pattern can be applied to any file extension.
+
+### Define Your Marker Type
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md:29]: Did you test those examples?
Because when it was raised seems like it was missing externalise something.
Is that working properly?
If so, could you please share in the description the POC / tests made?

**Comment by nerdeveloper** [comment]
## Testing Documentation

Hi @camilamacedo86! Thank you for the feedback. I've addressed all your review comments and wanted to share how I tested the implementation:

### Testing Approach

**1. Core Marker Implementation Testing**
I created standalone test programs to verify the marker generation logic:



**2. External Plugin API Compatibility**
Verified that the external plugin imports compile correctly:


**3. Documentation API Accuracy**
- ✅ Checked current external plugin patterns in 
- ✅ Updated examples to match the JSON/STDIN communication pattern used by current external plugins
- ✅ Verified import paths are correct for kubebuilder v4

### Changes Made Based on Review:
- ✅ Removed "Key Considerations" section per your feedback
- ✅ Updated external plugin example to use current API (JSON over STDIN/STDOUT)
- ✅ Simplified template section to be more practical
- ✅ Made documentation more concise and maintainable

The code examples are **minimal but functional** - they demonstrate the core pattern for creating custom markers that can be adapted for any file extension and integrated into external plugins following the current kubebuilder architecture.

**Comment by nerdeveloper** [comment]
## Testing Documentation

Hi @camilamacedo86! I've addressed all your review comments and tested the implementation thoroughly:

### Testing Results

**1. Core Marker Implementation**
- Tested marker creation for valid Rust files (.rs extension)
- Verified error handling for invalid file extensions  
- Confirmed formatPrefix function works with various inputs
- Template generation produces correct output with markers

**2. External Plugin API Compatibility**  
- Verified external plugin imports compile correctly with kubebuilder v4
- Confirmed PluginRequest and PluginResponse types are accessible
- Updated examples to match current JSON/STDIN communication pattern

**3. Code Examples Validation**
- All Go code examples compile without errors
- Marker generation follows the expected format: `// +rust:scaffold:imports`
- Error handling works correctly for unsupported file extensions
- Template generation produces valid Rust code with embedded markers

### Changes Made:
- Removed "Key Considerations" section per your feedback  
- Updated external plugin example to use current API (JSON over STDIN/STDOUT)
- Simplified template section to be more practical
- Made documentation more concise and maintainable
- Merged latest changes from master branch

The examples are minimal but functional, demonstrating the core pattern that can be adapted for any file extension.

**Comment by nerdeveloper** [comment]
## How to Test This Implementation

@camilamacedo86 Here are step-by-step instructions to test the custom marker implementation:

### Quick Test

**1. Test the core marker functionality:**
```bash
# Create a test file in any directory
cat > test_markers.go << 'EOF'
package main

import (
	"fmt"
	"path/filepath"
	"strings"
)

const RustPluginPrefix = "+rust:scaffold:"

type RustMarker struct {
	prefix  string
	comment string
	value   string
}

func NewRustMarker(path string, value string) (RustMarker, error) {
	ext := filepath.Ext(path)
	if ext != ".rs" {
		return RustMarker{}, fmt.Errorf("expected .rs file, got %s", ext)
	}
	return RustMarker{
		prefix:  formatPrefix(RustPluginPrefix),
		comment: "//",
		value:   value,
	}, nil
}

func (m RustMarker) String() string {
	return m.comment + " " + m.prefix + m.value
}

func formatPrefix(prefix string) string {
	trimmed := strings.TrimSpace(prefix)
	var builder strings.Builder
	if !strings.HasPrefix(trimmed, "+") {
		builder.WriteString("+")
	}
	builder.WriteString(trimmed)
	if !strings.HasSuffix(trimmed, ":") {
		builder.WriteString(":")
	}
	return builder.String()
}

func main() {
	fmt.Println("Testing marker creation...")
	marker, err := NewRustMarker("src/main.rs", "imports")
	if err != nil {
		fmt.Printf("Error: %v", err)
		return
	}
	fmt.Printf("Generated marker: %s", marker.String())
	
	// Test error case
	_, err = NewRustMarker("main.go", "imports")
	if err != nil {
		fmt.Printf("\nCorrect error for Go file: %v", err)
	}
}
EOF

# Run the test
go run test_markers.go
```

**Expected output:**
```
Testing marker creation...
Generated marker: // +rust:scaffold:imports
Correct error for Go file: expected .rs file, got .go
```

### Verify External Plugin Imports

**2. Test that external plugin types are accessible:**
```bash
cat > test_external.go << 'EOF'
package main

import (
	"fmt"
	"sigs.k8s.io/kubebuilder/v4/pkg/plugin/external"
)

func main() {
	var req external.PluginRequest
	var resp external.PluginResponse
	fmt.Printf("PluginRequest available: %T\n", req)
	fmt.Printf("PluginResponse available: %T\n", resp)
	fmt.Println("External plugin imports working!")
}
EOF

go run test_external.go
```

**Expected output:**
```
PluginRequest available: external.PluginRequest
PluginResponse available: external.PluginResponse
External plugin imports working!
```

### Clean Up
```bash
rm test_markers.go test_external.go
```

**Comment by nerdeveloper** [comment]
when all is good, i wil squash all of them into one commit

**Comment by camilamacedo86** [comment]
@nerdeveloper 

Could you please:
- Run `make remove-spaces` to fix the check. 
- And squash the commits I think it is good to get merged 

Thank you a lot for your contribution 🎉

**Comment by nerdeveloper** [comment]
@camilamacedo86 do you want me to also implement copilot suggestions on this pr.

**Comment by camilamacedo86** [comment]
Hi @nerdeveloper 

For we get this merged we need to:
- Run `make remove-spaces`
- Squash the commits
- And the reviews that make sense from copilot; can you please address?

**Comment by nerdeveloper** [comment]
@camilamacedo86 All requested changes have been completed:

✅ **Trailing spaces removed** - Ran `make remove-spaces` successfully
✅ **Commits squashed** - All 5 commits consolidated into 1 meaningful commit
✅ **Copilot suggestions addressed**:
  - Added note about `formatPrefix` being adapted from Kubebuilder's internal implementation
  - Added proper error handling for `json.Marshal` calls (2 instances)
  - Expanded `universe` comment to explain its purpose and usage

File: docs/book/src/plugins/extending/custom-markers.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    var builder strings.Builder
+    if !strings.HasPrefix(trimmed, "+") {
+        builder.WriteString("+")
+    }
+    builder.WriteString(trimmed)
+    if !strings.HasSuffix(trimmed, ":") {
+        builder.WriteString(":")
+    }
+    return builder.String()
+}
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply docs/book/src/plugins/extending/custom-markers.md:77]: @nerdeveloper I think it is a nice review

File: docs/book/src/plugins/extending/custom-markers.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+// Note: This formatPrefix implementation is adapted from Kubebuilder's internal
```
Comment (camilamacedo86) [inline docs/book/src/plugins/extending/custom-markers.md]: we have a class for note. 
Could you please check it out? 
Example; https://github.com/kubernetes-sigs/kubebuilder/blob/4c2adc8aaecd5055cdff61eec5ba3944105877cd/docs/book/src/migrations.md?plain=1#L49-L52

Also, remove from the block of code and link the Kubebuilder implementation?

File: docs/book/src/SUMMARY.md [hunk 11]
```diff
@@ -130,6 +130,7 @@
   - [Extending](./plugins/extending.md)
```
Comment (camilamacedo86) [inline docs/book/src/SUMMARY.md:128]: I noticed that the commit has been marked as invalid.
Could you please check if it contains an invalid character or something similar?
